src/preamble.js

```diff
@@ -796,6 +796,75 @@ function createExportWrapper(name, fixedasm) {
 }
 #endif
 
+#if ABORT_ON_WASM_EXCEPTIONS
+// When DISABLE_EXCEPTION_CATCHING != 1 `abortWrapperDepth` counts the recursion 
+// level of the wrapper function so that we only handle exceptions at the top level
+// letting the exception mechanics work uninterrupted at the inner level.
+// Additionally, `abortWrapperDepth` is also manually incremented in callMain so that 
+// we know to ignore exceptions from there since they're handled by callMain directly.
+var abortWrapperDepth = 0;
+
+// Instrument all the exported functions to:
+// - abort if an unhandled exception occurs
+// - throw an exception if someone tries to call them after the program has aborted
+// See settings.ABORT_ON_WASM_EXCEPTIONS for more info.
+function instrumentWasmExportsWithAbort(exports) {
+  var instExports = {};
+  for (var name in exports) {
+    (function(name) {
+      var original = exports[name];
+      
+      // Wrap all functions, copy the other symbols.
+      if (typeof original === 'function') {
+        instExports[name] = function() {
+          // Don't allow this function to be called if we're aborted!
+          if (ABORT == 1) { 
```

Comment:
isn't it also an error if `ABORT == 2`? why not check that as well?

Reply:
Before my changes `exit()` would set `abort=true` so any calls after that would fail. 

That's why I changed `var abort` from a bool to a number so that `abort()` sets it to 1 and `exit()` sets it to 2 and we can tell them apart. This was the smallest change I could make.

Here we only want to stop the function call if the program has previously aborted, but not when it exited.

Reply:
> Before my changes exit() would also set abort=true so any calls to the above after that would fail. But that's not what we want.

Why not? I think that would be correct. After calling `exit()` nothing else should be called, it's an error to do so. Throwing an error is the right behavior.

(In fact we should have assertions that verify that, with ASSERTIONS, so for the exit case the error below is redundant.)

Reply:
I'll check, but I think some of the test were failing which is why I made the change. I'll update here.

Reply:
Just re-ran all the tests and I remembered right - anything that uses `EXIT_RUNTIME=1`  and stuff like `printf` fails because there's this in library_syscall.js:

```
#if EXIT_RUNTIME == 1
  fd_write__postset: '__ATEXIT__.push(flush_NO_FILESYSTEM);',
#endif
```

Anything in __ATEXIT__ that runs a WASM function will fail, because `ABORT=true` would be set before running this in `exit()`:
```
ABORT = true;
EXITSTATUS = status;

exitRuntime(); ===> THIS CALLS __ATEXIT__ FUNCTIONS
```
It works fine if I move `ABORT=true` after `exitRuntime`, but I'm worried it will break something else, although doing a quick pass on parts of the code that use `ABORT` it seems safe to do so.
```
ABORT = true;
EXITSTATUS = status;

exitRuntime(); ===> THIS CALLS __ATEXIT__ FUNCTIONS
```
Doing this means I can avoid `ABORT=1/2` and also properly abort when stuff is called after `exit()`.

I'll re-run the tests now and we'll see...

Reply:
All tests pass with ABORT=true after exitRuntime() so I'll just go ahead and do that and revert the change of `var ABORT` to a number.

Reply:
It might be unrelated to this change but programs that abort should probably not call any __ATEXIT__ handlers.   I don't think any user code (including atexit handlers) should run once a program aborts.

Reply:
This was actually an `__ATEXIT__` handler from `library_syscall` to flush the streams. So it's library code and it's run _from within_ `exitRuntime()` :)

Reply:
Re-read your question and yeah, this was only for regular exit, not when the program aborts. That path doesn't execute any exit handlers.